Revamp the commit signal again and misc other fixes

whot opened the pull request.

whot committed:
Revert "Replace the resync signal with a commit-complete + cookie"

A better way to handle this is with a caller-supplied callback, there's
no need fo rthe signal to be a public one across all callers.

This reverts commit e2ae9d00df6c199e41b6d406f74da569fa29965b.

whot committed:
Revamp the commit approach again

Similar to the previous attempt, we generate a cookie and return it. But
we also take a callback that is passed through to the driver [1] and
called when it is done. This is a narrower approach than the previous
attempt at a commit-complete signal which would also notify other
listeners.

[1] we wrap that one to reset the dirty state but anyway...

whot committed:
tests: add a test for the roccat DPI change

This test should've prevented 4e4c9c9a895884d3d15f92eff739bd457b0e3c92

whot committed:
util: add type annotations

whot committed:
roccat: add type annotations

whot committed:
util: fix an incorrect exception type

whot committed:
roccat: remove unused variable

whot committed:
util: rename groupcount to repeat in the attrparser

A slightly less awkward name

whot committed:
Move the byte parser into a separate module

Slightly nicer code layout this way, can obsolete the other parser once
it's done.

whot committed:
hidpp20: add a few type annotations

whot committed:
hidpp20: switch two helpers to use attr

whot committed:
parser: add a pad_to keyword argument

For the cases where we have to parse into a fixed buffer.

whot committed:
parser: change the convert_to_data arguments into a proper class

This way we can add to it in the future without having to change every
caller, and callers don't have to care about the values they don't need.

whot committed:
hidpp20: rework the query handling for attr and our Parser

This simplifies the code a little and reduces the amount of init code we
need. All queries are now attr classes and the parsing is done via our
ratbag.parser.Parser (which gives us nice debugging).

whot committed:
hidpp20: switch profile parsing over to use attrs and Parser

whot committed:
hidpp20: remove unused send_and_recv_sync function

Leftover from some earlier version, no longer invoked and it'd cause an
exception anyway since it uses a variable that doesn't exist.

whot commented:
@amfern - I'm planning to merge this once the checks all pass, but if you are playing around with the hidpp20 code, please start from this branch. Sorry about all the mess with reworks, I'm using the holidays to learn a few new things :)

amfern commented:
This branch works great, I was able to list the profiles of my logitech g502

whot committed:
hidpp20: switch profile name/report_rate to properties

No need to store those values, let's keep the profile as close to the
wire profile as possible.

whot committed:
hidpp20: use the helper functions to connect to the recorder

whot committed:
hidpp20: add support for dpi list and report rate queries

whot committed:
hidpp20: fix the report rate to be an integer

whot committed:
Set up doctest

whot committed:
hidpp20: two more type annotations and a fixme

whot committed:
hidpp20: shut up mypy about an arg it doesn't know about

self._report_rates is set by the parser

whot committed:
tests: handle missing /dev/hidraw0 in the test

We can just glob this test over all hidraw devices, they all follow the
same rules anywya

whot committed:
parser: some doc updates

whot commented:
quickly merging this now that the CI finally passed (fdo is back up again, hopefully permanently)
